src/gt4py/next/program_processors/runners/gtfn_cpu.py

```diff
@@ -87,22 +101,25 @@ def decorated_program(*args):
 
             return decorated_program
 
-        otf_workflow: Final[workflow.Workflow[stages.ProgramCall, stages.CompiledProgram]] = (
-            gtfn_module.GTFNTranslationStep(
-                self.language_settings, self.enable_itir_transforms, self.use_imperative_backend
-            )
-            .chain(pybind.bind_source)
-            .chain(
-                compiler.Compiler(
-                    cache_strategy=cache.Strategy.SESSION, builder_factory=self.builder_factory
+        if cache_key not in self._cache:
+            otf_workflow: Final[workflow.Workflow[stages.ProgramCall, stages.CompiledProgram]] = (
+                gtfn_module.GTFNTranslationStep(
+                    self.language_settings, self.enable_itir_transforms, self.use_imperative_backend
                 )
+                .chain(pybind.bind_source)
+                .chain(
+                    compiler.Compiler(
+                        cache_strategy=cache.Strategy.SESSION, builder_factory=self.builder_factory
+                    )
+                )
+                .chain(convert_args)
             )
-            .chain(convert_args)
-        )
 
-        otf_closure = stages.ProgramCall(program, args, kwargs)
+            otf_closure = stages.ProgramCall(program, args, kwargs)
 
-        compiled_runner = otf_workflow(otf_closure)
+            compiled_runner = self._cache[cache_key] = otf_workflow(otf_closure)
```

Comment:
```suggestion
            compiled_runner = self._cache.setdefault(cache_key, otf_workflow(otf_closure))
```

The whole `if` clause could be avoided (since building the workflow costs almost nothing) by using this pattern. `setdefault(key, default)` returns the value associated with `key` if `key` is in the dict, otherwise it sets it to `default` *and* returns default, eliminating the need for `if cache_key in self._cache`.

Reply:
Maybe I'm missing something, but I don't see how the if else clause could be avoided. It is true that building the workflow is cheap (even though cheap, might be expensive from the perspective of running it on every stencil call), but what is not cheap is the call to the workflow `otf_workflow(otf_closure)`, which I only want to do in case the cache is not populated.

Reply:
Right, I overlooked that the workflow would be called anyway.